File: core-tests/e2e-tests/spring/spring-rpc-thrift/src/test/java/org/evomaster/e2etests/spring/rpc/examples/db/tree/DbTreeEMTest.java
package org.evomaster.e2etests.spring.rpc.examples.db.tree;

import com.foo.rpc.examples.spring.db.tree.DbTreeController;
import com.foo.rpc.examples.spring.db.tree.DbTreeService;
import com.foo.rpc.examples.spring.db.tree.DbTreeServiceImp;
import org.evomaster.core.problem.rpc.RPCCallResultCategory;
import org.evomaster.core.problem.rpc.RPCIndividual;
import org.evomaster.core.search.Solution;
import org.evomaster.e2etests.spring.rpc.examples.SpringRPCTestBase;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import java.util.Arrays;

import static org.junit.jupiter.api.Assertions.assertTrue;

public class DbTreeEMTest extends SpringRPCTestBase {

    @BeforeAll
    public static void initClass() throws Exception {

        SpringRPCTestBase.initClass(new DbTreeController());
    }

    @Test
    public void testRunEM() throws Throwable {

        runTestHandlingFlakyAndCompilation(
                "DbTreeEM",
                "org.bar.db.TreeEM",
                1000,
                (args) -> {


                    Solution<RPCIndividual> solution = initAndRun(args);

                    assertTrue(solution.getIndividuals().size() >= 1);

                    assertRPCEndpointResult(solution, DbTreeService.Iface.class.getName()+":get", RPCCallResultCategory.HANDLED.name());
                    assertAllContentInResponseForEndpoint(solution,DbTreeService.Iface.class.getName()+":get" ,
                            Arrays.asList(DbTreeServiceImp.NO_PARENT, DbTreeServiceImp.WITH_PARENT, DbTreeServiceImp.NOT_FOUND));
                });
    }
}

File: core-tests/e2e-tests/spring/spring-graphql/src/test/kotlin/org/evomaster/e2etests/spring/graphql/db/tree/DbTreeEMTest.kt
package org.evomaster.e2etests.spring.graphql.db.tree

import com.foo.graphql.db.tree.DbTreeController
import com.foo.graphql.db.tree.QueryResolver
import org.evomaster.core.EMConfig
import org.evomaster.core.problem.graphql.GQMethodType
import org.evomaster.e2etests.spring.graphql.SpringTestBase
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.ValueSource

class DbTreeEMTest : SpringTestBase() {

    companion object {
        @BeforeAll
        @JvmStatic
        fun init() {
            initClass(DbTreeController())
        }
    }

    @ParameterizedTest
    @ValueSource(booleans = [false, true])
    fun testEM(heuristicsForSQLAdvanced: Boolean){

        defaultSeed = 5

        runTestHandlingFlakyAndCompilation(
            "GQL_DbTreeEM",
            "org.foo.graphql.DbTreeEM" +
                    if (heuristicsForSQLAdvanced) "Complete" else "Partial",
            1000
        ) { args: MutableList<String> ->

            setOption(args, "problemType", EMConfig.ProblemType.GRAPHQL.toString())
            setOption(args, "heuristicsForSQL","true")
            setOption(args,"generateSqlDataWithSearch","true" )
            setOption(args,"heuristicsForSQLAdvanced", if (heuristicsForSQLAdvanced) "true" else "false")

            //issues with CI... trying to debug
            setOption(args,"useTimeInFeedbackSampling","false")

            val solution = initAndRun(args)

            Assertions.assertTrue(solution.individuals.size >= 1)
            assertHasAtLeastOne(solution, "dbTreeByParentId", GQMethodType.QUERY, 200, QueryResolver.NOT_FOUND)
            assertHasAtLeastOne(solution, "dbTreeByParentId", GQMethodType.QUERY, 200, QueryResolver.NO_PARENT)
            assertHasAtLeastOne(solution, "dbTreeByParentId", GQMethodType.QUERY, 200, QueryResolver.WITH_PARENT)
            /*
                with enabled taint analysis, there might exist errors while executing query
             */
            //assertNoneWithErrors(solution)
        }

    }
}
